src/bytes.rs

```diff
@@ -200,6 +201,53 @@ impl Bytes {
         }
     }
 
+    /// Create [Bytes] with a buffer whose lifetime is controlled
+    /// via an explicit owner.
+    ///
+    /// A common use case is to zero-copy construct from mapped memory.
+    ///
+    /// ```ignore
+    /// use bytes::Bytes;
+    /// use std::fs::File;
+    /// use memmap2::Map;
+    ///
+    /// let file = File::open("upload_bundle.tar.gz")?;
+    /// let mmap = unsafe { Mmap::map(&file) }?;
+    /// let b = Bytes::from_owner(mmap);
+    /// ```
+    ///
+    /// The `owner` will be transferred to the constructed [Bytes] object, which
+    /// will ensure it is dropped once all remaining clones of the constructed
+    /// object are dropped. The owner will then be responsible for dropping the
+    /// specified region of memory as part of its [Drop] implementation.
+    ///
+    /// Note that converting [Bytes] constructed from an owner into a [BytesMut]
+    /// will always create a deep copy of the buffer into newly allocated memory.
+    pub fn from_owner<T: AsRef<[u8]>>(owner: T) -> Self {
+        let owned = Box::into_raw(Box::new(Owned {
+            lifetime: OwnedLifetime {
+                ref_cnt: AtomicUsize::new(1),
+                drop: owned_box_and_drop::<T>,
+            },
+            owner,
+        }));
+
+        // Now that the ownership is moved to the Box its memory location is pinned.
+        // It's therefore safe to access the memory region, which will remain valid,
+        // even if the slice returned refers to memory within the owner object itself.
+        let owned_ref = unsafe { &(*owned) };
+        let buf = owned_ref.owner.as_ref();
```

Comment:
If `as_ref` were to panic we'd be leaking memory here because `Box::into_raw` has been called

Reply:
Good point. I'll fix this.

Reply:
Done.

Reply:
Miri is not liking it: `test owned_basic ... error: Undefined Behavior: trying to retag from <5921171> for SharedReadOnly permission at alloc2203252[0x18], but that tag does not exist in the borrow stack for this location`

A solution to this is probably to:

1. Go back to `Box::into_raw(Box::new(...))`
2. Put the pointer in a guard struct that on `Drop` does `Box::from_raw`
3. `mem::forget` the guard struct just before constructing `Bytes`

Reply:
Yeah, we must use `Box::into_raw` before calling `as_ref` for miri to be happy. You can reuse the destructor of `Bytes` to handle panics:
```rs
let owned = Box::into_raw(Box::new(Owned {
    lifetime: OwnedLifetime {
        ref_cnt: AtomicUsize::new(1),
        drop: owned_box_and_drop::<T>,
    },
    owner,
}));

let mut ret = Bytes {
    ptr: NonNull::dangling().as_ptr(),
    len: 0,
    data: AtomicPtr::new(owned.cast()),
    vtable: &OWNED_VTABLE,
};

let buf = unsafe { &*owned }.owner.as_ref();
ret.ptr = buf.as_ptr();
ret.len = buf.len();

ret
```

Reply:
Please add a comment explaining this so that refactors don't break it in the future. Please also add a test that the value is dropped on panic.